### ISSUE TITLE
Incorrect compilation of `await` when the value is a `FutureOr<Future<...>>`

### ISSUE THREAD

**Issue by osa1** [dart-lang/sdk#52484]
Language spec section 17.34 (in my own words) says that in an `await x` where `x`'s compile-time type is `S` and runtime type is `R`, the `await`ed value depends on whether `R` is a subtype of `Future<flatten(S)>`. (`flatten` removes one layer of `Future` and `FutureOr`, e.g. `flatten(Future<X>)` is `X`)

- If `R <: Future<flatten(S)>` then we await on the value
  (`Future<flatten(S)>` replaces the outer `FutureOr` with `Future`)

- Otherwise we await on `Future.value(o)`

For example, when `R` is a `Future<int>` but `S` is `FutureOr<Future<int>>` we need to await on `Future.value(future)` and not `future`. Repro:

```dart
import "dart:async";

Future<int> f1() => Future<int>.value(1);

FutureOr<Future<int>> f2() => Future<int>.value(1);

void main() async {
  print(await f1());
  print(await f2());
}
```

Here the second line should print something like `Instance of Future<int>`, but we currently await on the value and print `1`.

Since runtime type of a value should always be subtype of the compile-time type of the value (for type soundness), I think what these rules boil down to is `FutureOr<Future<X>>` should not be treated as a `Future<X>` but as `Future<Future<X>>`, by passing the value to `Future.value` before awaiting.

Relevant test: language/await_type_test

**Comment by eernstg** [comment]
This is probably not the same issue, but I noticed recently that the test 'await_flatten_test.dart' fails with every CFE configuration, because (apparently) it computes a wrong result for `flatten(T)` for one special form of `T`. Cf. https://github.com/dart-lang/sdk/issues/52485. Just thought you might want to check whether there's a connection. ;-)

**Comment by lrhn** [comment]
Also be careful around `return e;` in `async` methods, which *may* do an implicit `await` on the value. 
The context type is the future-value-type, `F`, of the surrounding future, so the context type of `e` is `FutureOr<F>`.
If the static type of `e` is-a `Future<F>`, it should be awaited, otherwise it should not (it's not wrapped in a `Future` like an `await`, although since we don't promise anything about the timing of returns from `async` functions, it can probably safely do so).
That's not the same as `await` where the check depends on the static type of the expression (through flatten), but instead on the context type. (I really hope no code depends on that difference. It's bad enough as it is.)

**Comment by osa1** [comment]
> Also be careful around return e; in async methods, which may do an implicit await on the value.
> ...

@lrhn I think this is the last paragraph in spec 18.12, right? I tried this:

```dart
import "dart:async";

Future<int> f1() => Future<int>.value(1);

FutureOr<int> f2() async {
  Future<int> x = f1();
  return x;
}

void main() async {
  print(f2());
}
```

Here in `return x`, the function is `async`, runtime type of `x` is a subtype of `Future<flatten(FutureOr<int>)>` (= `Future<int>`), but `f2` returns the future, without awaiting, in VM and dart2js. Is this a bug? If I'm reading the spec right the `return` here should await.

**Comment by eernstg** [comment]
The execution of `f2()` should return a future which has been created for that asynchronous invocation, and the future which is the value of `x` at `return x;` should be awaited at that point. If it is completed with an object (here: `1`) then that object should be used to complete the returned future. Dart-pad behaves like this, and prints 'false' when the two futures are checked for identity. Could this be what you are seeing?

```dart
import "dart:async";

Future<int> f1() => Future<int>.value(1);

FutureOr<int> f2() async {
  Future<int> x = xInF2 = f1();
  return x;
}

late Future<int> xInF2;

void main() async {
  print(identical(f2(), xInF2)); // 'false'.
}
```

**Comment by osa1** [comment]
@eernstg Ah, that's indeed what I'm seeing. Thanks for the example.

**Comment by osa1** [comment]
This was fixed with f23abcd5931476396c8e3c8fed8240fb3b6186f0.
